packages/flutter_tools/lib/src/artifacts.dart

```diff
@@ -62,6 +63,9 @@ enum Artifact {
   /// Tools related to subsetting or icon font files.
   fontSubset,
   constFinder,
+
+  /// The location of file generators.
+  flutterToolsFileGenerators,
```

Comment:
The JS files are also artifacts that need to be served by the web server, or go to the build output of a flutter app. They just happen to be checked-in to the repo.

In a previous iteration of this, I didn't need to use `Artifacts` because I was able to rely on [`Cache.defaultFlutterRoot(...)`](https://github.com/flutter/flutter/blob/973cff40b4022edd59dbf44475ffc569bece1ea8/packages/flutter_tools/lib/src/cache.dart#L228-L228). But that didn't work in Google3.

Then Jonah suggested I look into [`Google3Artifacts`](https://source.corp.google.com/piper///depot/google3/mobile/flutter/cli/lib/context.dart;l=641?pli=1) since that gives us a way to override the paths when running inside Google3.

I'm not super familiar with the integration with Google3, so I'm open to ideas and suggestions on how to make this work.

Reply:
> (or rather Cache.flutterRoot which everyone else seems to use in fluter_tools)

Can't use `Cache.flutterRoot` because it defaults to `/` or empty string in tests. I want this path to always point to the real files regardless of the environment.

> Without considering google3, would you prefer using `Artifacts` or `Cache.defaultFlutterRoot`

I'm not sure if we prefer `Artifacts` or `Cache.flutterToolsRoot`. I personally prefer the `Cache.flutterToolsRoot` solution. Maybe @jonahwilliams has an opinion here?

> In this case specifically, I would recommend creating a new `Cache.flutterToolsRoot` which defaults to `path.join(Cache.flutterRoot, 'packages', 'flutter_tools')`

I like this idea a lot. It is consistent with how `Cache.flutterRoot` works today.

> We could even go one step further to update everywhere else that uses `${Cache.flutterRoot}/packages/flutter_tools` to use `Cache.flutterToolsRoot` directly

Yep, this makes sense. There are quite a bit of callsites that need to be updated. What do you think about making the `Cache.flutterToolsRoot` change in its own PR to keep it clean?

Reply:
Cache.whatever is almost certainly a static field that has to be carefully initialized during startup or tests. i would avoid using it at all costs.

Reply:
> Cache.whatever is almost certainly a static field that has to be carefully initialized during startup or tests. i would avoid using it at all costs.

I can confirm. I tried to do this now with careful initialization and still got a ton of test failures.

How about something similar to [`getFlutterRoot()`](https://github.com/flutter/flutter/blob/492d57262fafba0d38bad47bc5dff3d7ada77ac3/packages/flutter_tools/test/src/common.dart#L42-L42)? Can we override global functions like that in Google3?

Reply:
I don't fully know the context for @chingjun's answer but I disgree pretty heavily there. The API for looking up paths in the tool is artifacts. It should be irrelevant that a particular artifact does or does not live in the cache. local engine artifacts don't live in the cache, for example

Reply:
@jonahwilliams so you are in favor of the `Artifacts` approach if I'm understanding correctly?

Reply:
Yes

Reply:
Hey thanks for adding me. Sorry, I don't know enough about flutter tools to have an opinion here so I'll leave Chingjun to represent google3 :-D.

Reply:
I realised Chingjun is OOO for a while so I'm happy to help out in google3 if necessary. @mdebbar let me know if you need any help figuring out the g3fix. I don't know much about artifacts, but I can learn!

Reply:
I commented on the g3fix CL but forgot to comment here. I'm going with the `Artifacts` approach if there are no objections.